src/plugins/ui_actions/public/triggers/trigger_internal.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Trigger, TriggerContext, TriggerContextParams } from './trigger';
+import { TriggerContract } from './trigger_contract';
+import { UiActionsService } from '../service';
+import { Action } from '../actions';
+import { buildContextMenuForActions, openContextMenu } from '../context_menu';
+
+/**
+ * Internal representation of a trigger kept for consumption only internally
+ * within `ui_actions` plugin.
+ */
+export class TriggerInternal<T extends Trigger<any, any>> {
+  public readonly contract = new TriggerContract<T>(this);
+
+  public get id() {
+    return this.trigger.id;
+  }
+
+  constructor(public readonly service: UiActionsService, public readonly trigger: T) {}
+
+  public createContext(params: TriggerContextParams<T>): TriggerContext<T> {
+    if (!this.trigger.createContext) return (params as unknown) as TriggerContext<T>;
+    return this.trigger.createContext(params);
```

Comment:
What use cases do you see this using different parameter input vs context output?

I think the same thing could be achieved by using an action that mutates the context and executes the next trigger.  I think having two possible types to this system might unnecessarily over complicate things? Unless I am missing a use case where this would be handy.

Reply:
My thinking was to provide easy and type-safe way of creating *context* objects.

- easy: `params` can be a subset of the things needed to create the `context` and the trigger can construct the context.
- type-safe: it provides a way to pass in context into `Trigger` interface, so we can use it for type checking.

Reply:
We had some sync discussion, decided to remove `createContext`.